aten/src/ATen/native/cpu/FusedAdamKernel.cpp

```diff
@@ -0,0 +1,147 @@
+#include <ATen/core/Tensor.h>
+#include <ATen/Parallel.h>
+#include <ATen/native/DispatchStub.h>
+#include <ATen/native/fused_adam.h>
+#include <ATen/Dispatch.h>
+#include <ATen/cpu/vec/vec.h>
+namespace at::native {
+
+namespace{
+
+template <typename scalar_t>
+void adam_fused_step_impl(
+    const at::Tensor& param,
+    const at::Tensor& grad,
+    const at::Tensor& exp_avg,
+    const at::Tensor& exp_avg_sq,
+    const at::Tensor& max_exp_avg_sq,
+    const at::Tensor& state_step,
+    const double lr,
+    const double beta1,
+    const double beta2,
+    const double weight_decay,
+    const double eps,
+    const bool amsgrad,
+    const bool maximize,
+    const float* grad_scale_ptr) {
+  int64_t step = state_step.item<int64_t>();
+  scalar_t* param_data = param.data_ptr<scalar_t>();
+  scalar_t* exp_avg_data = exp_avg.data_ptr<scalar_t>();
+  scalar_t* exp_avg_sq_data = exp_avg_sq.data_ptr<scalar_t>();
+  scalar_t* max_exp_avg_sq_data = amsgrad ? max_exp_avg_sq.data_ptr<scalar_t>() : nullptr;
+  scalar_t* grad_data = grad.data_ptr<scalar_t>();
+
+  // need to use double here to align with non-fused adam
+  double bias_correction1 = 1 - std::pow(beta1, step);
+  double step_size =  lr / bias_correction1;
+  double bias_correction2 = 1 - std::pow(beta2, step);
+  double exp_avg_grad_coefficient = 1 - beta1;
+  double exp_avg_sq_grad_coefficient = 1 - beta2;
+  double bias_correction2_sqrt = std::sqrt(bias_correction2);
+
+  using Vec = at::vec::Vectorized<scalar_t>;
+
+  // update momentum vt and mt
+  // also accumulate sum of param_norm and rtw_norm
+  at::parallel_for(
+      0, param.numel(), 0, [&](int64_t begin, int64_t end) {
+        // local pointers
+        scalar_t* param_ptr = param_data + begin;
```

Comment:
I guess we should make the data cache line aligned at least.

Reply:
Hi, @jgong5.
The `param_data` is the first memory address of a Tensor and I suppose it should be cache line aligned while do `cpu_malloc`. So the `grain_size=0` will use defalut `grain_size` which is 32K. Then the `param_data + begin` should also be cache line aligned.

Reply:
I guess `grain_size` determines the number of threads needed for the parallel region but the `begin` is decided by the chunk size which is not equivalent to the `grain_size`. So `grain_size` is 32k doesn't mean `param_data + begin` is cache aligned.

Reply:
I updated logics to make sure `param_data + begin` is cache line aligned for `main loop`. There mighe be a `tail loop` for openmp backend.